### PR TITLE
TRUNK-6202 Replace Hibernate Criteria API with JPA for HibernateConceptDAO

### PR DESCRIPTION
**Pull request by k4pran**
## Description of what I changed

This task is to migrate the deprecated hibernate criteria to jpa criteria for the `HibernateConceptDAO` class.

Additionally, I added some additional tests where I noticed gaps, I validated these tests with both the original hibernate code and the migrated JPA code.

There are a few changes I want to highlight where it wasn't a as direct translation from hibernate to JPA:

- I noticed some of the case matching was not consistent, for example [here](https://github.com/openmrs/openmrs-core/compare/master...k4pran:TRUNK-6202-HibernateConceptDAO?expand=1#diff-913f109df0354bd959a725513fa00929c58e113728e1c1f4bb7f920b06b89609L353) the `isDatabaseStringComparisonCaseSensitive` seemed the wrong way around. 

- In JPA you can't order by a non-selected field in combination with `distinct`. In these cases I filtered by distinct outside of the query. An example of this can be seen [here](https://github.com/openmrs/openmrs-core/compare/master...k4pran:TRUNK-6202-HibernateConceptDAO?expand=1#diff-913f109df0354bd959a725513fa00929c58e113728e1c1f4bb7f920b06b89609R1123)

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6202

## Checklist: I completed these to help reviewers :)

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/api/ConceptServiceTest.java [hunk 1]
```diff
@@ -185,7 +186,201 @@ public void getConceptByName_shouldGetConceptByPartialName() {
...
+			}
+			previousProposal = proposal;
+		}
+	}
+
+	/**
+	 * @see ConceptService#getConceptProposals(String) )
+	 */
+	@Test
+	public void getConceptProposals_shouldReturnProposalsMatchingTextAndUnmappedState() {
```
Comment (tendomart) [inline api/src/test/java/org/openmrs/api/ConceptServiceTest.java:277]: There are some unnecessary spaces in this method

File: api/src/test/java/org/openmrs/api/ConceptServiceTest.java [hunk 2]
```diff
@@ -185,7 +186,201 @@ public void getConceptByName_shouldGetConceptByPartialName() {
...
+				assertEquals(actualConcept.getConceptId(), proposal.getMappedConcept().getConceptId());
+			}
+		}
+	}
+
+	/**
+	 * @see ConceptService#getProposedConcepts(String) )
+	 */
+	@Test
+	public void getProposedConcepts_shouldReturnEmptyWhenNoMappedConcepts() {
```
Comment (tendomart) [inline api/src/test/java/org/openmrs/api/ConceptServiceTest.java:324]: Extra spaces

File: api/src/test/java/org/openmrs/api/ConceptServiceTest.java [hunk 3]
```diff
@@ -185,7 +186,201 @@ public void getConceptByName_shouldGetConceptByPartialName() {
...
+			assertNotNull(next.getSortWeight());
+			assertTrue(current.getSortWeight() <= next.getSortWeight());
+		}
+	}
 	
+	/**
+	 * @see ConceptService#getAllConceptProposals(boolean)
+	 */
+	@Test
+	public void getAllConceptProposals_WhenIncludeCompletedIsTrue_shouldReturnAllProposals() {
```
Comment (tendomart) [inline api/src/test/java/org/openmrs/api/ConceptServiceTest.java:360]: extra spaces

File: api/src/test/java/org/openmrs/api/ConceptServiceTest.java (hunk 1 above)
Comment (k4pran) [reply api/src/test/java/org/openmrs/api/ConceptServiceTest.java:277]: Removed

**Comment by dkayiwa** [comment]
@k4pran did you find it uncomfortable creating the smaller pull requests? 😊

File: api/src/test/java/org/openmrs/api/ConceptServiceTest.java [hunk 4]
```diff
@@ -412,7 +607,68 @@ public void saveConcept_shouldKeepIdForNewConceptIfOneIsSpecified() {
...
+				foundRetired = true;
+			} else {
+				foundNonRetired = true;
+			}
+			if (foundRetired && foundNonRetired) {
+				break;
+			}
+		}
+
+		assertTrue(foundRetired, "No retired concept classes found.");
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/ConceptServiceTest.java:668]: Shouldn't the message say the opposite?

**Comment by k4pran** [comment]
> @k4pran did you find it uncomfortable creating the smaller pull requests? 😊

Haha 😂. The plan was to do one DAO at a time to break them up but I guess this was a big DAO, I can try to break some of the larger DAOs into smaller PRs in the future but not sure if there is any natural way to divide them up except a few methods at a time?

**Comment by dkayiwa** [comment]
> I can try to break some of the larger DAOs into smaller PRs in the future but not sure if there is any natural way to divide them up except a few methods at a time?

No need to break larger DAOs. Let us go with what you have already done of one DAO at a time. 😊

**Pull request merged**